### ISSUE TITLE
Peer discovery implementation

### ISSUE THREAD

**Issue by larteyoh** [larteyoh/testshop#6]
```
Details
```
I am looking for a way for peers to find other peers running the neroshop daemon. The protocol used must be scalable and efficient.

**If I have to resort to using a blockchain or libp2p to achieve this goal then I do not mind at all, as long as I can bring this project to life. I don't care what networking library you use. Asio, LibZMQ, LibUV, etc. Anything works!**

_I am exploring methods and protocols that could be used to create a peer-to-peer network_
**Possible candidates include:**
- [Gossip protocol](https://github.com/izeigerman/pittacus)
- [Distributed Hash Tables (DHT)](https://github.com/jech/dht)
- hybrid - an approach that combines both DHTs and Gossip protocols and may be the most effective way to achieve fault tolerance.

Both DHTs and Gossip protocols are highly scalable and can scale to the billions. I believe they both are also fault tolerant.

```
Related files
```
https://github.com/larteyoh/testshop/blob/main/src/core/server.cpp
https://github.com/larteyoh/testshop/blob/main/src/core/client.cpp
https://github.com/larteyoh/testshop/tree/main/src/daemon/main.cpp
https://github.com/larteyoh/testshop/blob/main/src/core/protocol/*

```
References
```
https://en.bitcoin.it/wiki/Satoshi_Client_Node_Discovery
https://www.techopedia.com/definition/25772/peer-discovery (ELI5)
https://stackoverflow.com/questions/310607/peer-to-peer-methods-of-finding-peers
https://github.com/mitrapinaki/PeerToPeer
https://github.com/libredrop/peer-discovery
https://github.com/peteristhegreat/qt-udp-discovery

```
References - P2P Network basics
```
https://developer.bitcoin.org/devguide/p2p_network.html

```
References - DNS seed nodes
```
https://www.reddit.com/r/Bitcoin/comments/4dicny/dns_seeds_servers_hardcoded_in_bitcoin/
https://bitcoin.stackexchange.com/questions/14371/what-is-a-dns-seed-node-vs-a-seed-node
https://bitcoin.stackexchange.com/questions/116931/how-does-one-become-a-dns-seed-for-bitcoin-core
https://github.com/bitcoin/bitcoin/blob/1b2460bd5824170ab85757e35f81197199cce9d6/src/chainparams.cpp#L112
https://github.com/sipa/bitcoin-seeder

```
Videos
```
gossip protocol:
- https://www.youtube.com/watch?v=sPAYvBnIkqo

```
Documentations
```
libuv:
- http://docs.libuv.org/en/v1.x/ (Official docs)
- https://nikhilm.github.io/uvbook/networking.html
- https://github.com/wqvbjhc/libuv\_tcp
- https://gist.github.com/leiless/0a33314cd5ea046a5841ac3287620642
- https://github.com/manishmanu/libuv\_pipe\_server\_client

libzmq:
- https://zguide.zeromq.org
- http://api.zeromq.org

```
Related issues
```
[Rework client and server code by integrating any cross-platform C/C++ networking library](https://github.com/larteyoh/testshop/issues/47)

[Tor and I2P integration](https://github.com/larteyoh/testshop/issues/7)

```
Obstacles in P2P networking
```
Outbound-only nodes are nodes that can establish outgoing connections to other nodes in the network, but they have limited or no ability to accept incoming connections. They can send requests, exchange information, and participate in the network by initiating outbound connections, but they cannot receive incoming connections from other nodes.

Outbound-only nodes are often used by users who have restricted network configurations, such as port forwarding disabled or UPnP disabled. These nodes rely on seed nodes or other nodes with incoming connectivity to bootstrap their initial connections and join the network.

While outbound-only nodes can still actively participate in the network by initiating outbound connections, it's important to note that their limited reachability can result in fewer incoming connections compared to nodes with full incoming and outgoing connectivity. This may have a minor impact on the node's ability to receive updates or propagate information to a wide range of peers.

And to combat, this issue, I have strictly made the default binding IP address to 0.0.0.0, though it is not recommended as it can expose the node to potential security risks when it comes to public-facing interfaces. This will allow Neroshop nodes to listen on all available network interfaces, both local and public-facing. This configuration enables the client to accept incoming connections from other peers in the Neroshop network.

```
Bounty reward 
```
$0

**Comment by vandal-prog** [comment]
I'm interested

**Comment by vandal-prog** [comment]
How do I reach out to you?

**Comment by larteyoh** [comment]
I have a matrix room where we discuss the development of neroshop [here](https://matrix.to/#/#neroshop:matrix.org). I can also be contacted via email at larteyoh@protonmail.com.
